dataplane/api-server/src/netutils.rs

```diff
@@ -5,59 +5,89 @@ SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
-/// Linux systems with iproute2 installed.
-///
-/// TODO: replace this https://github.com/Kong/blixt/issues/49
-pub fn if_name_for_routing_ip(ip_addr: Ipv4Addr) -> Result<String, Error> {
-    // run the linux command "ip route" to get the device responsible for
-    // routing the given IP address.
-    let ip = ip_addr.to_string();
-    let mut cmd = Command::new("ip");
-    let child = cmd
-        .arg("route")
-        .arg("get")
-        .arg("to")
-        .arg(&ip)
-        .stdout(Stdio::piped())
-        .spawn()?;
-
-    // grab the output from the "ip route" command
-    let output = child.wait_with_output()?;
-    let stdout = from_utf8(output.stdout.as_slice())?;
-
-    // construct a regex to match the output
-    let mut regex_str = ip.clone();
-    regex_str.push_str(r" (via [0-9]+\.[0-9]+\.[0-9]+\.[0-9]+ )?dev ([a-zA-Z0-9]+)\s+");
-    let re = Regex::new(&regex_str)?;
-
-    // match on the output to find the network device responsible for routing
-    // the provided IP address.
-    let match_err = format!("no device found to route {}", ip);
-    let device = re
-        .captures(stdout)
-        .ok_or_else(|| Error::msg(match_err.clone()))?
-        .iter()
-        .last()
-        .ok_or_else(|| Error::msg(match_err.clone()))?
-        .ok_or_else(|| Error::msg(match_err))?
-        .as_str()
-        .to_owned();
-
-    Ok(device)
+// pub fn if_nametoindex(ifname: String) -> Result<u32, Error> {
+//     let ifname_c = CString::new(ifname)?;
+//     let ifindex = unsafe { libc_if_nametoindex(ifname_c.into_raw()) };
+//     Ok(ifindex)
+// }
+
+/// Returns an net interface index for a Ipv4 address, like `ip route get to`
+pub fn if_index_for_routing_ip(ip_addr: Ipv4Addr) -> Result<u32, Error> {
+    // run the linux command "ip route" to get the device's index responsible for
+    // routing the given Ipv4 address.
+    let mut socket = Socket::new(NETLINK_ROUTE).unwrap();
+    let _port_number = socket.bind_auto().unwrap().port_number();
+    socket.connect(&SocketAddr::new(0, 0)).unwrap();
+
+    let mut nl_hdr = NetlinkHeader::default();
+    nl_hdr.flags = NLM_F_REQUEST | NLM_F_DUMP_FILTERED;
+
+    // construct RouteMessage
+    let route_header = RouteHeader {
+        address_family: AddressFamily::Inet,
+        flags: RouteFlags::LookupTable,
+        destination_prefix_length: 32,
+        table: RouteHeader::RT_TABLE_MAIN,
+        ..Default::default()
+    };
+    let route_attribute = RouteAttribute::Destination(RouteAddress::Inet(ip_addr));
+    let mut route_message = RouteMessage::default();
+    route_message.attributes = vec![route_attribute];
+    route_message.header = route_header;
+
+    let no_device_err: String = format!("no device found to route {}", ip_addr);
+    let con_packet_err: String = "construct packet failed".to_string();
+    let nl_send_msg_err: String = "netlink send message failed".to_string();
+    let nl_recv_msg_err: String = "netlink receive message failed".to_string();
```

Comment:
While it will probably be a bit more code, it may be ultimately more idiomatic to provide a custom error type for these conditions? Actually it occurs to me, does the netlink library not have custom errors that cover these conditions? :thinking:

Reply:
When I modified this, I realized that there are only two remaining error types. It seems a bit excessive to create a custom error type just for these two. (sorry for making so many sloppy mistakes, and thank you for your patience.)

Reply:
No worries at all. Have you confirmed whether netlink has some types already in place to cover these error conditions? If they don't and we just have the two, then it's probably fine to keep these as simple error message strings for now and we can maybe even improve this in the netlink library later.

Reply:
yep,I don't find some error type in netlink to cover these two.